core/offchain/src/api.rs

```diff
@@ -44,25 +44,144 @@ enum ExtMessage {
 
 /// A persisted key seed.
 #[derive(Encode, Decode)]
-struct CryptoKey {
+struct StoredKey {
 	kind: CryptoKind,
 	phrase: String,
 }
 
-enum Key {
-	Sr25519(sr25519::Pair),
+impl StoredKey {
+	fn generate_with_phrase(kind: CryptoKind, password: Option<&str>) -> Self {
+		match kind {
+			CryptoKind::Ed25519 => {
+				let phrase = ed25519::Pair::generate_with_phrase(password).1;
+				Self { kind, phrase }
+			}
+			CryptoKind::Sr25519 => {
+				let phrase = sr25519::Pair::generate_with_phrase(password).1;
+				Self { kind, phrase }
+			}
+		}
+	}
+
+	fn to_local_key(&self, password: Option<&str>) -> Result<LocalKey, ()> {
+		match self.kind {
+			CryptoKind::Ed25519 => {
+				ed25519::Pair::from_phrase(&self.phrase, password)
+					.map(|x| LocalKey::Ed25519(x.0))
+			}
+			CryptoKind::Sr25519 => {
+				sr25519::Pair::from_phrase(&self.phrase, password)
+					.map(|x| LocalKey::Sr25519(x.0))
+			}
+		}
+		.map_err(|e| {
+			warn!("Error recovering Offchain Worker key. Password invalid? {:?}", e);
+			()
+		})
+	}
+}
+
+enum LocalKey {
```

Comment:
I'd merge the whole enum with `Key` for simplicity to be honest, but it's a really minor grumble